research/delf/delf/python/whiten.py

```diff
@@ -0,0 +1,119 @@
+      application.
+  """
+  eps = 1e-6
+  # If `dimensions` is not specified, we do not perform dimensionality
+  # reduction.
+  if not dimensions:
+    dimensions = projection.shape[0]
+
+  X = np.dot(projection[:dimensions, :], X - mean_descriptor_vector)
+  X_whitened = X / (np.linalg.norm(X, ord=2, axis=0, keepdims=True) + eps)
+
+  return X_whitened
+
+
+def whitenlearn(X, qidxs, pidxs):
+  """Learning the post-processing of fine-tuned descriptor vectors.
+
+  This method of whitening learning leverages the provided labeled data and
+  uses linear discriminant projections. The projection is decomposed into two
+  parts: whitening and rotation. The whitening part is the inverse of the
+  square-root of the intraclass (matching pairs) covariance matrix. The
+  rotation part is the PCA of the interclass (non-matching pairs) covariance
+  matrix in the whitened space. The described approach acts as a
+  post-processing step, equivalently, once the fine-tuning of the CNN is
+  finished. For more information about the method refer to the section 3.4
+  of https://arxiv.org/pdf/1711.02512.pdf.
+
+  Args:
+    X: List of L2-normalized descriptors.
+    qidxs: List of query indexes.
+    pidxs: List of positive pairs indexes.
+
+  Returns:
+    mean_descriptor_vector: Mean descriptor vector.
+    projection: Whitening projection matrix.
+  """
+  # Calculating the mean descriptor vector, which is used to perform centering.
+  mean_descriptor_vector = X[:, qidxs].mean(axis=1, keepdims=True)
+  # Interclass (matching pairs) difference.
+  interclass_difference = X[:, qidxs] - X[:, pidxs]
+  covariance_matrix = np.dot(interclass_difference, interclass_difference.T) / \
+                      interclass_difference.shape[1]
+
+  # Whitening part.
+  projection = np.linalg.inv(cholesky(covariance_matrix))
+
+  projected_X = np.dot(projection, X - mean_descriptor_vector)
+  non_matching_covariance_matrix = np.dot(projected_X, projected_X.T)
+  eigval, eigvec = np.linalg.eig(non_matching_covariance_matrix)
+  order = eigval.argsort()[::-1]
+  eigvec = eigvec[:, order]
+
+  # Rotational part.
+  projection = np.dot(eigvec.T, projection)
+  return mean_descriptor_vector, projection
+
+
+def cholesky(matrix):
+  """Cholesky decomposition.
+
+  Cholesky decomposition suitable for non-positive definite matrices: involves
+  adding a small value `alpha` on the matrix diagonal until the matrix
+  becomes positive definite.
+
+  Args:
+    matrix: Square matrix.
+
+  Returns:
+    L: Upper-triangular Cholesky factor of `matrix`, a matrix with real and
+      positive diagonal entries.
+  """
+  alpha = 0
+  while True:
+    try:
+      # If the input parameter matrix is not positive-definite,
+      # the decomposition fails and we iteratively add a small value `alpha` on
+      # the matrix diagonal.
+      L = np.linalg.cholesky(matrix + alpha * np.eye(*matrix.shape))
+      return L
+    except:
```

Comment:
please follow the rules for exceptions as specified at: https://google.github.io/styleguide/pyguide.html#24-exceptions

In particular, "Never use catch-all except: statements, or catch Exception or StandardError, unless you are (..)"

So here you should basically specify the exception.